Issue 5287- Endpoints to get the md5 hash of the original model file

Bogdan-Crn opened the pull request:
This fixes #5287 

Description

- Created endpoint for containers to allow retrieval of the MD5 data for the container
- Created endpoint for federations to allow retrieval of the MD5 data for all the containers in the federation that the user has access to
- Created a new function in FileRefs to update a specific .ref file, this is to allow the caching of MD5 hashes for future reference

Test cases

- Unit tests for containers getting the MD5 data, first time the database should not have the MD5Hash property and the second time the system should have set it in the db
- Unit tests for federation getting the MD5 data, similar to above 
- Unit test for the FileRefs new function
- E2E tests for containers and federations testing scenarios where the user has different levels of rights

carmenfan commented on backend/tests/v5/unit/processors/teamspaces/projects/models/federations.test.js:
```diff
@@ -359,6 +367,97 @@ const testGetTicketGroupById = () => {
+			return mockConatiners.filter((container) => container._id === containers[0]);
+		});
+
+		// it should get an empty array if user doesn't have rights to the container
```
redundant comments should be removed 😆 that's just the test description

Bogdan-Crn commented:
Moved most of the modelList logic to the filesManager.
Updated the federation check to see if the user has access to the container.
Updated what happens if the federation is empty.
Updated the tests to contain generateRandomString() where necessary.

carmenfan commented on backend/src/v5/processors/teamspaces/projects/models/commons/modelList.js:
```diff
@@ -65,68 +62,35 @@ ModelList.getModelList = async (teamspace, project, user, modelSettings) => {
 
-	const code = UUIDParse.unparse(rev._id.buffer);
+	const code = rev.tag;
 	const uploadedAt = new Date(rev.timestamp).getTime();
```
Leave this as a Date object for now and do the serialisation in the middleware

carmenfan commented on backend/src/v5/processors/teamspaces/projects/models/commons/modelList.js:
```diff
@@ -65,68 +62,35 @@ ModelList.getModelList = async (teamspace, project, user, modelSettings) => {
+	if (!rev.rFile?.length) return {};
 
-	const code = UUIDParse.unparse(rev._id.buffer);
+	const code = rev.tag;
```
Sorry I probably wasn't being clear last time 😅 

Can we keep this as tag on the return value?

Bogdan-Crn replied:
Ok, so instead of code the key will be named tag, what about the value that it holds? Should that stay as the tag of the revision?

carmenfan replied:
yea same data just call it tag on the response is good 👍

carmenfan commented on backend/src/v5/services/filesManager.js:
```diff
@@ -214,4 +216,25 @@ FilesManager.storeFileStream = async (teamspace, collection, id, dataStream, met
+FilesManager.getMD5FileHash = async (teamspace, container, filename) => {
+	const refEntry = await getRefEntry(teamspace, `${container}.history.ref`, filename);
+
+	if (Object.keys(refEntry).includes('MD5Hash')) {
```
I think you can probably abbreviate this bit with ES6 style syntax?

Also can we call it `md5Hash` just for consistency in naming convention?

```suggestion
const {size, md5Hash: hash }= await getRefEntry(teamspace, `${container}.history.ref`, filename);

if(hash) {
  return {size, hash};
}
```

carmenfan commented on backend/src/v5/services/filesManager.js:
```diff
@@ -214,4 +216,25 @@ FilesManager.storeFileStream = async (teamspace, collection, id, dataStream, met
+		};
+	}
+
+	const file = await FilesManager.getFile(teamspace, `${container}.history`, filename);
```
getting filesManager to append history is a bit too coupled (whilst we're only using it in history atm, we can in theory get md5 for anything).

I would rename container to collection and get the processor to pass in `${container}.history` instead

carmenfan reviewed (changes requested):
Looking good, just a few minor comments 😆
